### PR TITLE
Incorrect order status label in sales update emails sent from the admin for non-default store views issue29263

### LINKED ISSUE #29263: Incorrect order status label in sales update emails sent from the admin for non-default store views
After sending a sales update email from the admin for an order using a non-default store view, the resulting email contains the order status label for the default store view instead of the order status label for the store view in which the order is placed.

### Preconditions (*)

1. A Magento 2.4-develop installation with sample data.

### Steps to reproduce (*)
1. In the admin:
    * Go to “Stores > All Stores”, click on “Create Store View”, fill in the required values, set the status to “Enabled” and click on the "Save Store View” button.
    * Go to “Stores > Order Status”, click on the “Pending” status, enter the following data under “Store View Specific Labels”, and click on the “Save Status” button.
        * Default Store View: “Pending (default)”
        * New Store View: “Pending (secondary)”
Screenshot from 2020-07-29 12-26-46

2. In the frontend:
    * Select the new store view in the store view switcher.
    * Add a product to the cart.
    * Complete the checkout using the “Check / Money order” payment method.
3. In the admin:
    * Go to “Sales > Orders”, and click on the newly created order.
    * On this order view page, enter a comment, click on the “Notify Customer by Email” checkbox and click on the "Submit Comment” button.

### Expected result (*)

1. An order update email is sent containing “Pending (secondary)” as the order status.

### Actual result (*)

1. An order update email is sent containing “Pending (default)” as the order status.

Screenshot of the resulting email:

sales_order_update_email_with_incorrect_order_status_label

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by edenduong**
### Description (*)

Fix issue below:

### Fixed Issues (if relevant)

1. Fixes magento/magento2#29263: Incorrect order status label in sales update emails sent from the admin for non-default store views 

### Manual testing scenarios (*)

1. In the admin:
    * Go to “Stores > All Stores”, click on “Create Store View”, fill in the required values, set the status to “Enabled” and click on the "Save Store View” button.
    * Go to “Stores > Order Status”, click on the “Pending” status, enter the following data under “Store View Specific Labels”, and click on the “Save Status” button.
        * Default Store View: “Pending (default)”
        * New Store View: “Pending (secondary)”
2. In the frontend:
    * Select the new store view in the store view switcher.
    * Add a product to the cart.
    * Complete the checkout using the “Check / Money order” payment method.
3. In the admin:
    * Go to “Sales > Orders”, and click on the newly created order.
    * On this order view page, enter a comment, click on the “Notify Customer by Email” checkbox and click on the "Submit Comment” button.

**=> An order update email is sent containing “Pending (secondary)” as the order status.**

### Questions or comments

I changed public function getStatusFrontendLabel(?string $code) to public function getStatusFrontendLabel(?string $code, int $storeId = null). We need the $storeId params. Please consider the Semantic Version Checker.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by edenduong** [comment]
@magento run all tests

**Comment by edenduong** [comment]
I  changed `public function getStatusFrontendLabel(?string $code)`  to `public function getStatusFrontendLabel(?string $code, int $storeId = null)`. We need the $storeId params. Please consider the Semantic Version Checker.

**Comment by rogyar** [comment]
Hi @edenduong. Thank you for the fix. 
Changing the signature of `getStatusFrontendLabel` might be considered as a backward-incompatible change. However, the fix looks really useful, sending it for further approval.

**Comment by aadmathijssen** [comment]
Hi @edenduong, thanks for the PR! I think this should fix the issue I reported.

There still seems to be another related issue with the `getStoreLabel` method however: for the case that no store specific status label exists for the `$storeId` parameter, the standard `__` translation function is used; this function ignores the value of `$storeId`. In order to fix that, we could use environment emulation when `$storeId` is not the current store id.

**Comment by swnsma** [comment]
Hi @edenduong!
The 3'rd param is really looks useful.
However, can we implement it in backward-compatible way (e.g. introduce new method with 3 params and deprecate old one with 2)?

File: app/code/Magento/Sales/Model/Order/Config.php [hunk 1]
```diff
@@ -156,12 +159,13 @@ public function getStatusLabel($code)
      * Retrieve status label for area
      *
      * @param string|null $code
+     * @param int|null $storeId
      * @return string|null
      * @since 102.0.1
      */
-    public function getStatusFrontendLabel(?string $code): ?string
+    public function getStatusFrontendLabel(?string $code, int $storeId = null): ?string
```
Comment (sidolov) [inline app/code/Magento/Sales/Model/Order/Config.php]: Unfortunately, we cannot change the signature of the public method in a class marked as @api.

**Review by sidolov — CHANGES_REQUESTED** [review]
Please, use suggested way how to avoid backward-incompatible changes described in [comment](https://github.com/magento/magento2/pull/29273#issuecomment-671766629)

**Comment by edenduong** [comment]
Thank you. I think I should create new public method and @deprecated the old method.

File: app/code/Magento/Sales/Model/Order.php [hunk 2]
```diff
@@ -3,6 +3,8 @@
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
  */
+declare(strict_types=1);
```
Comment (engcom-Echo) [inline app/code/Magento/Sales/Model/Order.php]: This is correct but there is a big risk. For example, "public function getCustomerName ()" will be fail. I think this change doesn't relate to this PR.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Review by gabrieldagama — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8254](https://jira.corp.magento.com/browse/ENGCOM-8254) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. In the admin:
    * Go to “Stores > All Stores”, click on “Create Store View”, fill in the required values, set the status to “Enabled” and click on the "Save Store View” button.
    * Go to “Stores > Order Status”, click on the “Pending” status, enter the following data under “Store View Specific Labels”, and click on the “Save Status” button.
        * Default Store View: “Pending (default)”
        * New Store View: “Pending (second)”
2. In the frontend:
    * Select the new store view in the store view switcher.
    * Add a product to the cart.
    * Complete the checkout using the “Check / Money order” payment method.
3. In the admin:
    * Go to “Sales > Orders”, and click on the newly created order.
    * On this order view page, enter a comment, click on the “Notify Customer by Email” checkbox and click on the "Submit Comment” button.

**Before:** :heavy_multiplication_x: An order update email is sent containing “Pending (default)” as the order status.



**After:** :heavy_check_mark: An order update email is sent containing “Pending (second)” as the order status.



Also was checked with different types of statuses, on different websites, without specific labels, and everything works as expected.

**Comment by engcom-Echo** [comment]
Need approval. Added a new public method instead of the old.

**Comment by sivaschenko** [comment]
Approval request: https://jira.corp.magento.com/browse/MC-38232

**Review by swnsma — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-8254](https://jira.corp.magento.com/browse/ENGCOM-8254) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

The result is the same as in the [comment](https://github.com/magento/magento2/pull/29273#issuecomment-698172943).

**Comment by engcom-Charlie** [comment]
Needs to approve since the public method has been added to `Magento/Sales/Model/Order/Config`.

**Comment by sivaschenko** [comment]
@edenduong @engcom-Charlie adding the new public method for the `Magento/Sales/Model/Order/Config` was not approved. OMS functionality is highly customizable area and the chance that someone extended the Config class is very high.

Can we extract logic related to status labels to a separate class and use in the Order model?

**Comment by swnsma** [comment]
Hi @sivaschenko,
As far as I can see, the customisation was done in backward-compatible way. 

@edenduong, would you share your plans about future of the PR?

**Comment by engcom-Alfa** [comment]
Hi @edenduong , Thanks for your contribution and collaboration!

As per the linked issue's [last comment](https://github.com/magento/magento2/issues/29263#issuecomment-847561365), the fix is already done. You may cross check bug fix in latest Magento `2.4-develop` version, in case of any more issues, feel free to report bugs/PR.
Thanks!

**Pull request closed without merging**